src/vcall.cpp

```diff
@@ -1326,7 +1326,12 @@ void jitc_var_vcall_collect_data(tsl::robin_map<uint64_t, uint32_t, UInt64Hasher
     if (!it_and_status.second)
         return;
 
-    const Variable *v = jitc_var(index);
+    Variable *v = jitc_var(index);
+    ThreadState *ts = thread_state(v->backend);
+
+    /* Scalar literals created outside of this virtual function call should be
+       considered as data to avoid baking them inside of the kernel IR. */
+    bool ext_literal = v->literal && v->cse_scope != ts->cse_scope;
```

Comment:
We also change CSE scope in other situations (e.g. loop recording). Maybe this is too broad? Perhaps it's better store an index somewhere that records the location of the innermost function being recorded right now. Anything older than that is opaque-ified.

Reply:
I am now using the `ts->vcall_self_index` variable which is created right before tracing the virtual function, hence it can act as a lower bound for the variable index to detect external scalar literals in vcalls.

Reply:
great!